### PR TITLE
dryrun tool

### PR DESCRIPTION
**Pull request by haibinxie**
1. add reconcile API call to storageCluster to populate default values and generate other objects
2. write one object to one file instead of all objects to a file
3. add validations to migration
4. restructure code

### CODE REVIEW

File: pkg/dryrun/dryrun.go [hunk 1]
```diff
@@ -0,0 +1,810 @@
...
+	if err = d.writeFiles(objs); err != nil {
+		return err
+	}
+	logrus.Infof("Operator will deploy %d objects, saved to folder %s", len(objs), d.outputFolder)
+
+	if d.realClient != nil {
+		controller := &storagecluster.Controller{}
+		controller.SetKubernetesClient(d.realClient)
+		h := migration.New(controller)
+		dsObjs, err := h.GetAllDaemonSetObjects(d.cluster.Namespace)
```
Comment (dsupure) [inline pkg/dryrun/dryrun.go:228]: Are you assuming we only have portworx daemonset in the namespace. Is it possible you have multiple DaemonSets ?

File: pkg/dryrun/dryrun.go [hunk 2]
```diff
@@ -0,0 +1,810 @@
...
+					kind = "StatefulSet"
+					logrus.Info("px-csi-ext is deployed as StatefulSet instead of Deployment by operator")
+				}
+			}
+		} else if kind == "Prometheus" && name == "prometheus" {
+			logrus.Warning("Prometheus service name will change from prometheus to px-prometheus after migration")
+			name = "px-prometheus"
+		} else if kind == "Service" && name == "prometheus" {
+			name = "px-prometheus"
+		} else if kind == "Service" && name == "autopilot" {
```
Comment (dsupure) [inline pkg/dryrun/dryrun.go:464]: Do we need to consider ccm here?

File: pkg/dryrun/dryrun.go [hunk 3]
```diff
@@ -0,0 +1,810 @@
...
+	}
+
+	start := func() {
+		if err := d.realControllerMgr.Start(ctrl.SetupSignalHandler()); err != nil {
+			logrus.Fatalf("Manager exited non-zero error: %v", err)
+		}
+	}
+	go start()
+	// Wait for the cache to be started.
+	time.Sleep(time.Second)
```
Comment (dsupure) [inline pkg/dryrun/dryrun.go:639]: Is 1 sec enough?

File: pkg/dryrun/validation.go [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+	"fmt"
+	"reflect"
+	"sort"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/libopenstorage/operator/pkg/util"
+)
+
+var (
```
Comment (dsupure) [inline pkg/dryrun/validation.go:13]: Why we need to define twice in seperate files? Can we have one place instead?

File: cmd/dryrun/dryrun.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+func main() {
+	app := cli.NewApp()
+	app.Name = "portworx-operator-dryrun"
+	app.Usage = "Portworx operator dry run tool"
+	app.Version = version.Version
+	app.Action = execute
+
+	app.Flags = []cli.Flag{
+		cli.BoolFlag{
```
[the remark this answers is not in the document]
Comment (haibinxie) [reply cmd/dryrun/dryrun.go:28]: good idea, resolved, there is no aliases param, i used a different way.

File: cmd/dryrun/dryrun.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+			Name:  flagVerbose,
+			Usage: "Enable verbose logging",
+		},
+		cli.StringFlag{
+			Name:  flagStorageCluster,
+			Usage: "File for storage cluster spec, retrieve from k8s if it's not configured",
+		},
+		cli.StringFlag{
+			Name:  flagKubeConfig,
```
[the remark this answers is not in the document]
Comment (haibinxie) [reply cmd/dryrun/dryrun.go:37]: what do you mean?

File: pkg/dryrun/dryrun.go [hunk 7]
```diff
@@ -0,0 +1,810 @@
...
+	mockDriver   storage.Driver
+	cluster      *corev1.StorageCluster
+	outputFolder string
+}
+
+// Init performs required initialization
+func (d *DryRun) Init(kubeconfig, outputFolder, storageClusterFile string) error {
+	var err error
+
+	k8sVersion := "v1.22.0"
```
[the remark this answers is not in the document]
Comment (haibinxie) [reply pkg/dryrun/dryrun.go:102]: if kubeconfig is not provided, 1.22.0 will be the default version to setup simulator.

File: pkg/dryrun/dryrun.go [hunk 8]
```diff
@@ -0,0 +1,810 @@
...
+	} else {
+		ver, err := testutil.GetK8SVersion()
+		if err != nil {
+			logrus.WithError(err).Error("failed to get k8s version")
+		} else {
+			k8sVersion = ver
+			logrus.Infof("k8s version is %s", k8sVersion)
+			minVer, err := goversion.NewVersion("1.16")
+			if err != nil {
+				return fmt.Errorf("error parsing version '1.16': %v", err)
```
[the remark this answers is not in the document]
Comment (haibinxie) [reply pkg/dryrun/dryrun.go:115]: yes, right now it's actually blocked during migration too

File: pkg/dryrun/dryrun.go (hunk 1 above)
Comment (haibinxie) [reply pkg/dryrun/dryrun.go:228]: yes, there is a ticket to support two daemonset installs, it's not supported yet. we will document this prior to release

File: pkg/dryrun/dryrun.go (hunk 2 above)
Comment (haibinxie) [reply pkg/dryrun/dryrun.go:464]: ccm is part of telemetry. we dont support in daemonset spec however

File: pkg/dryrun/dryrun.go (hunk 3 above)
Comment (haibinxie) [reply pkg/dryrun/dryrun.go:639]: yes, this should be good. it just needs to start without the cache to be fully populated.

File: pkg/dryrun/validation.go (hunk 4 above)
Comment (haibinxie) [reply pkg/dryrun/validation.go:13]: var is only used by the current file

**Comment by haibinxie** [comment]
@dsupure resolved comments in a new commit, also updated the confluence page with the two docs about min k8s version and two daemonset installs.

File: cmd/dryrun/dryrun.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (dsupure) [reply cmd/dryrun/dryrun.go]: I mean keep the name simple.  can we use `kubeconfig` instead of `flagKubeConfig`.

**Review by dsupure — APPROVED** [review]
LGTM!

**Pull request merged**